packages/bpk-component-infinite-scroll/src/DataSource.js

```diff
@@ -0,0 +1,87 @@
+/*
+ * Backpack - Skyscanner's Design System
+ *
+ * Copyright 2018 Skyscanner Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* @flow */
+
+class DataSource {
+  listeners: Array<() => void>;
+
+  onDataChange: (callback: (updatedElements: Array<any>) => void) => boolean;
+
+  offDataChange: (callback: (updatedElements: Array<any>) => void) => boolean;
+
+  triggerListeners: (...args: Array<any>) => void;
+
+  constructor() {
+    this.listeners = [];
+  }
+
+  /* eslint-disable-next-line no-unused-vars, class-methods-use-this */
+  fetchItems(index: number, nElements: number): Promise<Array<any>> {
+    throw new Error('Not implemented');
+  }
+
+  onDataChange(callback: () => void): boolean {
+    if (this.listeners.indexOf(callback) === -1) {
+      this.listeners.push(callback);
+      return true;
+    }
+    return false;
+  }
+
+  offDataChange(callback: () => void): boolean {
+    const index = this.listeners.indexOf(callback);
+    if (index !== -1) {
+      this.listeners.splice(index, 1);
+      return true;
+    }
+    return false;
+  }
+
+  triggerListeners(...args: Array<any>) {
+    this.listeners.forEach(cb => cb(...args));
+  }
+}
+
+export class ArrayDataSource extends DataSource {
```

Comment:
This should be `export class ArrayDataSource<T=any> extends DataSource<T>` with the generic treatment

Reply:
done